File: 05_basic_recursion/01_palindrome/solution.py
# https://leetcode.com/problems/valid-palindrome/
class Solution:
    def isPalindrome(self, s: str) -> bool:
        cleanStr = ''.join(ch.lower() for ch in s if ch.isalnum())
        return self.isPalindromeRecursive(cleanStr, 0, len(cleanStr)-1)

    def isPalindromeRecursive(self, s: str, left: int, right: int):
        if left >= right:
            return True

        if s[left] != s[right]:
            return False

        return self.isPalindromeRecursive(s, left+1, right-1)

File: 05_basic_recursion/01_palindrome/solution.java
// https://leetcode.com/problems/valid-palindrome/
class Solution {
    public boolean isPalindrome(String s) {
        String cleanStr = s.toLowerCase().replaceAll("[^a-z0-9]", "");
        return isPalindromeRecursive(cleanStr, 0, cleanStr.length()-1);
    }

    public boolean isPalindromeRecursive(String s, int left, int right) {
        if (left >= right) {
            return true;
        }
        
        if (s.charAt(left) != s.charAt(right)) {
            return false;
        }

        return isPalindromeRecursive(s, left+1, right-1);
    }
}